### ISSUE TITLE
Quick Copy regex issue

### ISSUE THREAD

**Issue by aj3423** [aj3423/SpamBlocker#150]
The question is, how to get the number from this SMS content:

> Your verification number is 12345678. Please enter this and press the # key.

Solutions:
- `is.{0,2}(\d+)` , I'd prefer this one, simple and generic.
- `is (\d+)`,  restrict to `is` followed by 1 space.
- `(?<=is )\d{8}`, lookarounds are complex, not recommended in most cases.
- `(?<=is.{0,2})\d{8}`,  more generic, it would also allow `is: 12345678` or `is:12345678`

**Comment by noto10** [comment]
> The question is, how to get the number from this SMS content:
> 
> > Your verification number is 12345678. Please enter this and press the # key.

I've been tweaking some regex to cover a multitude of code SMSes. Still refining the process to get a good generic one which covers most, if not all encountered cases.
 
> Solutions:
> 
>     * `is.{0,2}(\d+)` , I'd prefer this one, simple and generic.

I'm going to use this solution as a basis for my question:

Perhaps I don't understand the part outlined in red:

<img width="252" alt="copy1" src="https://github.com/user-attachments/assets/8b4f46fe-315a-42ca-a6f4-89c502bdbfd6">

 but when I tried using

`is.{0,2}(\d+)`

I don't see any **Copy: xxxxxxxx** button in either the phone's native SMS notification or Spamblocker's notification as seen below:

<img width="324" alt="copy2" src="https://github.com/user-attachments/assets/3cccf51b-5aa2-4ad5-a8f7-10c8e63c33da">

<img width="324" alt="copy3" src="https://github.com/user-attachments/assets/15b2b695-f107-42a7-9228-02ba22f15d8b">

<img width="324" alt="copy4" src="https://github.com/user-attachments/assets/5bac7109-89be-473a-bbc3-e5a1808b0de9">

Where is the button and what am I missing?

**Comment by aj3423** [comment]
![Screenshot](https://github.com/user-attachments/assets/ec4e762b-76c2-40d4-8233-f964dc22c4f3)

This notification doesn't seem to come from this app, it's shown by the app "Enterprise Hub". The correct icon is supposed to be:

![Screenshot2](https://github.com/user-attachments/assets/8c0befb9-ff2d-493c-af4d-018468158133)

Maybe this message didn't pass the rule check, you can check the history log in the SMS tab to see why it's blocked.

After solving this, you can disable the notification of the Enterprise Hub to prevent double notifications.

**Comment by noto10** [comment]
> ![Screenshot](https://private-user-images.githubusercontent.com/4710875/375805357-ec4e762b-76c2-40d4-8233-f964dc22c4f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg3MDA3MzIsIm5iZiI6MTcyODcwMDQzMiwicGF0aCI6Ii80NzEwODc1LzM3NTgwNTM1Ny1lYzRlNzYyYi03NmMyLTQwZDQtODIzMy1mOTY0ZGMyMmM0ZjMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTJUMDIzMzUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDc1ODUzNzlkNmI0ZGI3M2ZjNWFlYzdkNjExZjRjMmYxZTY3ZDc2NjQyZWViZGM5ZDNhMzc0ZGEzNzdjZDhmZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.GM8d55tsPDqatl8ipyaC7AYnxJDS9pBgDYidjhzqXVQ)
> 
> This notification doesn't seem to come from this app, it's shown by the app "Enterprise Hub". 

The above notification is from the phone's native SMS app. There's a notification for every incoming SMS. 'Enterprise Hub' is just the sender in this case, not any manager. If the sender were a contact named 'John', it will reflect 'John' instead of 'Enterprise Hub', with a 'J' in the icon, like this:

![image](https://github.com/user-attachments/assets/e4dd6a7c-e61c-4c69-ad2d-54997be5ad91)


>The correct icon is supposed to be: 
>
> ![Screenshot2](https://private-user-images.githubusercontent.com/4710875/375805999-8c0befb9-ff2d-493c-af4d-018468158133.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg3MDA3MzIsIm5iZiI6MTcyODcwMDQzMiwicGF0aCI6Ii80NzEwODc1LzM3NTgwNTk5OS04YzBiZWZiOS1mZjJkLTQ5M2MtYWY0ZC0wMTg0NjgxNTgxMzMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTJUMDIzMzUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmUxMjE2NzU1NjczM2E0ZGVlY2JlN2UwZDU3YjczN2Q1NDllNDUyM2IxNmRhNWQ3ZTk0MDNlMmRmYTU0MWNmMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.MrJgRCEANaeovB9nTk97SidNPqVnhytPkFPb_UOTGLQ)

I do also have the Spamblocker notification, which is a separate one as indicated in my previous post:

![image](https://github.com/user-attachments/assets/c85aa89b-250d-4da1-8251-5c11bd8380c1)

But there is no Copy button.

**Comment by aj3423** [comment]
When creating a quick copy rule, only the  "for passed" is checked by default, so the button will only appear for passed messages. Your message in the image is a blocked one, so there is no button. 
![image](https://github.com/user-attachments/assets/0d3e030d-723e-4b10-bf7f-7ebb1ecc60bd)
You can test with a passed message, or make the rule also work "for blocked".

**Comment by noto10** [comment]
> When creating a quick copy rule, only the "for passed" is checked by default, so the button > will only appear for passed messages. Your message in the image is a blocked one, so there > is no button.

> ![image](https://private-user-images.githubusercontent.com/4710875/375919124-0d3e030d-723e-4b10-bf7f-7ebb1ecc60bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg3MjIwMTQsIm5iZiI6MTcyODcyMTcxNCwicGF0aCI6Ii80NzEwODc1LzM3NTkxOTEyNC0wZDNlMDMwZC03MjNlLTRiMTAtYmY3Zi03ZWJiMWVjYzYwYmQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTJUMDgyODM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjM0NjI5OGE3ZDJhNzM0NDU4NjdjNjk4YTE1M2QyYjUwODFhMWQ3MTE1NzZhNzJmZGIzYTdlZDRmNDhlNzhkMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.zricfoJ3SPbSNH64m4buS9sIvq8p8WGgoDGOZdo_zgQ)

>  You can test with a passed message, or make the rule also work "for blocked".

Tested and works.

<img width="270" alt="copy7" src="https://github.com/user-attachments/assets/7bb77c16-c1ed-45ac-af0c-4f134a72ee06">
<br>
<br>


Thank you for clarifying about the button only appearing for "passed" messages.

Questions:

1. Priority:

From my understanding, priority value is assigned arbitrarily by the user and the rule with higher priority is always processed before another with lower priority. What about rules with equal priorities? What is the order for processing them?

2.  Quick Copy vs Message Rules 

Is there any kind of overlap? Quick Copy appears to be an implementation of regex to a special subset of Messages.

**Comment by aj3423** [comment]
1. Rules with same priority  
The checking order would depend on its id in the database, which means which one was created first, it's not a reliable. 
**Blacklist and whitelist rules should always have different priorities.** All whitelist rules have priority as 10 by default, and 0 for blacklist rules. Rules of the same type can have the same prioritiy, because when it passes, it doesn't matter which rule passes it.

2. Quick Copy / Message Rule
One for checking whether it should be blocked, the other one for extracting content,  making the logic more easier to understand. Maybe technically it's possible to combine them, but it would cause more confusion and the regex could be very hard to write. One more thing is that, the quick copy also supports copying phone number.

**Comment by noto10** [comment]
>        **Blacklist and whitelist rules should always have different priorities.** All whitelist rules have priority as 10 by default, and 0 for blacklist rules. Rules of the same type can have the same prioritiy, because when it passes, it doesn't matter which rule passes it.

I don't see 10 for whitelist default and 0 for blacklist default though. Whichever one I choose in the configuration popup has 1 as the default.

> 
>     2. Message Rule /Quick Copy
>        One for checking whether it should be blocked, the other one for extracting content,  making the logic more easier to understand. Maybe technically it's possible to combine them, but it would cause more confusion and the regex could be very hard to write. One more thing is that, the quick copy also supports copying phone numbers.

After thinking about it, I would agree that integrating Quick Copy into Message Rule would make it a tad too complex to navigate.

**Comment by aj3423** [comment]
> I don't see 10 for whitelist default and 0 for blacklist default though. 

It's not in the tooltip or readme. The quick setting options have priorities of either 10 or 0, 10 for allowing, 0 for blocking(only three of them):
- contact exclusive 
- stir exclusive 
- spam database

About the default priority 1 of regex rule, actually 0 would also work. But the "type" is "blacklist" by default, if you want to change it to "whitelist", then it will conflict with the above three quick setting options, you would have to manually change the priority to a value higher than 0, that's why I choose 1 by default.

**Comment by noto10** [comment]
>  About the default priority 1 of regex rule, actually 0 would also work. But the "type" is "blacklist" by default, if you want to change it to "whitelist", then it will conflict with the above three quick setting options, you would have to manually change the priority to a value higher than 0, that's why I choose 1 by default.

I experienced some hours of strange regex failure earlier when a few regex's I was trying failed to Copy code correctly (code was truncated with the last digit missing) when implemented in Spamblocker but various regex checkers confirmed the validity of the regex's. This kept happening until I force stopped Spamblocker and reopened it again. After that, the same regex's performed correctly. I have not been able to reproduce this behavior and as such, I hesitate to report this as a bug. Just FYI and I'll keep an eye out if it should happen again.

At the moment, I have more than 10 entries under Regex Settings > Number Rules and nearly as many under Message Rules. This makes the whole Settings rather long to scroll through. Would it be possible to make it an option to use a collapse/expand subsection to make it easier to manage?

**Comment by aj3423** [comment]
Is that message still there in the SMS app? Try to copy the content and test it in the Testing dialog, make sure to use copy/paste  instead of manually typing, because such problems are usually caused by special characters. 
And how long is that message? Longer than 256 characters?

I'll add this: click the rule header to collapse/expand the list.

**Comment by aj3423** [comment]
collapse/expand is done: https://github.com/aj3423/SpamBlocker/actions/runs/11359492937

**Comment by aj3423** [comment]
> Could you display the number of rules like in the database?

it's not displayed directly on the UI, there isn't much space left, but you can see the rule count in the context menu, the "Delete All(30) Rules".

**Comment by noto10** [comment]
> Is that message still there in the SMS app? Try to copy the content and test it in the Testing dialog, make sure to use copy/paste instead of manually typing, because such problems are usually caused by special characters. And how long is that message? Longer than 256 characters?

The message in question was much less than 256 characters and I did use copy/paste since I was testing so many regex's with different messages. I did make sure to pay attention to things like no unintentional trailing spaces/tabs. I also made sure that I clicked Save after edit of the regex. The only thing I didn't do was to exit/kill Spamblocker so I could switch back and forth from the SMS app quickly. Is there any possibility that Spamblocker was reading from a cached value of the regex despite my clicking on Save? That would explain why it only started working after the force stop.

In any case, it was a one-off incident, albeit a very lengthy episode that stumped me. I'll report if it should happen again.

> 
> I'll add this: click the rule header to collapse/expand the list.

Have installed and it looks great. Thank you.
